### PR TITLE
Stop console and Docker logging on shutdown

### PR DESCRIPTION
**Pull request by JamesNK**
Left over changes from https://github.com/dotnet/aspire/pull/1578

* DashboardServiceData dispose now stops any outstanding console and Docker container logging.
* Unit test console logging
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1620)

### CODE REVIEW

**Comment by IEvangelist** [comment]
Any chance you could include an update in this PR that removes the unnecessary starting of Docker Desktop, when a .NET Aspire app host starts — that doesn't contain any containers?

File: src/Aspire.Hosting/Dashboard/DockerContainerLogSource.cs [hunk 1]
```diff
@@ -1,17 +1,30 @@
...
+    private readonly string _containerId;
+    private readonly CancellationToken _cancellationToken;
+
+    public DockerContainerLogSource(string containerId, CancellationToken cancellationToken)
+    {
+        _containerId = containerId;
+        _cancellationToken = cancellationToken;
+    }
+
     public async IAsyncEnumerator<IReadOnlyList<(string Content, bool IsErrorMessage)>> GetAsyncEnumerator(CancellationToken cancellationToken = default)
```
Comment (adamint) [inline src/Aspire.Hosting/Dashboard/DockerContainerLogSource.cs:23]: Is this going to be a different token than _cancellationToken?

**Review by adamint — APPROVED** [review]

**Review by drewnoakes — COMMENTED** [review]
What exactly is the issue that's been addressed here? I can't see any behaviour change, and there's a bunch of non-functional, subjective style changes and the addition of disposable here, which I called out on your previous PR as creating confusion between dispose/cancellation.

The PR title suggests that logging doesn't stop on shutdown, but that doesn't match my experience.

**Comment by JamesNK** [comment]
This PR is replicating what the resource publisher and other subscriptions do with having a centralized token for cancellation that is triggered when the app is stopping, and a token passed in by the caller. There is no way for app shutdown to explicitly force logging to stop.

For example, in publisher:

https://github.com/dotnet/aspire/blob/5ee685ee232856f25de5ae59ba0325e2d8e11448/src/Aspire.Hosting/Dashboard/ResourcePublisher.cs#L47-L49

It looks different in console/Docker logging because logging has classes that implement `IAsyncEnumerable` instead of a method that returns `IAsyncEnumerable`, e.g.

```cs
public async IAsyncEnumerator<IReadOnlyList<(string Content, bool IsErrorMessage)>> GetAsyncEnumerator(CancellationToken cancellationToken = default)
{
    using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(_cancellationToken, cancellationToken);
    var linkedToken = linkedCts.Token;

    // ...
}
```

But the concept is exactly the same.

**Comment by drewnoakes** [comment]
I think a clear problem statement would help. If I understand correctly, your concern is that the processes that pull logs from disk and from docker might not be cleaned up when the application shuts down.

If that's the case, I think https://github.com/dotnet/aspire/pull/1633 achieves the same thing without changing the design. When the application is shut down, any enumeration of console logs is cancelled. That PR uses cancellation token rather than `IDisposable`, which I think makes things clearer.

**Comment by JamesNK** [comment]
This PR creates consistency of log subscriptions with resource subscriptions. Resource subscriptions support being centrally cleaned up when the app stops. Currently log subscriptions don't.

This isn't fixing an immediate bug but what is currently happening relies on every caller of logs to correctly cancel them. If they don't then the executing code isn't cleaned up. That's especially bad for Docker logging because even if the process exists, it could leave a docker process running.

> rather than `IDisposable`

This uses cancellation:

```cs
public async IAsyncEnumerator<IReadOnlyList<(string Content, bool IsErrorMessage)>> GetAsyncEnumerator(CancellationToken cancellationToken = default)
{
    using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(_cancellationToken, cancellationToken);
    var linkedToken = linkedCts.Token;

    // ...
}
```

It works exactly like the resource publisher.

**Comment by JamesNK** [comment]
What about this? [`35a6772` (#1620)](https://github.com/dotnet/aspire/pull/1620/commits/35a67721d550b368b1c482f7775d371ac7870474)

**Comment by drewnoakes** [comment]
I don't believe this PR actually changes anything (since #1633). The application shutting down token is now used, which will cancel all pending async enumerations, which will tear down any log streaming. This adds a bunch of code without changing behaviour, unless I'm missing something.

**Comment by JamesNK** [comment]
The changed behavior is if someone incorrectly doesn't cancel a console stream, then this ensures console logging logic - which includes starting a process - are canceled by the time the host shuts down.

You cannot guarantee that everyone who will ever call the API in the future does the right thing. This puts up a guard rail.

**Comment by drewnoakes** [comment]
Cancellation during shutdown was added in #1633. It doesn't require users to do the right thing. The enumeration is gated on the app hosts shutting down token, so if either the app shuts down, or the client cancels the operation, things are cleaned up. I can't see a way to leak things in the code today.

In more detail, here's what I believe happens during shutdown. Let me know if I'm missing anything.

1. `ShuttingDown` token is cancelled
2. All pending stream subscriptions are terminated
3. DI instances are disposed

This PR hooks cleanup in that step 3, but everything's already cleaned up by that point (modulo race conditions).

**Comment by JamesNK** [comment]
> It doesn't require users to do the right thing.

It requires everyone who uses console log publisher correctly cancel. A week ago that wasn't happening. At this moment they are. In 1 or 2 years someone who doesn't know could pick up the console log publisher, not do the right thing and then `docker` processes could start leaking again.

It's a guard rail against misuse.

**Comment by drewnoakes** [comment]
Requiring cancellation vs. requiring disposal doesn't seem that different to me. Both require external users to do the right thing. These are internal components, so we know all usages of the code.

I can see value in adding a check that the passed cancellation token is actually cancellable (added in #1658). But offering disposal in addition to cancellation doesn't add any value that I can see, and as I commented on the first PR you offered this change in, I feel it makes things more confusing and verbose. Sorry, but I'm not in favour of this change.

**Comment by JamesNK** [comment]
I think it is confusing that the [resource publisher does this](https://github.com/dotnet/aspire/blob/5ee685ee232856f25de5ae59ba0325e2d8e11448/src/Aspire.Hosting/Dashboard/ResourcePublisher.cs#L47-L49), but the log publisher doesn't.

Looks like we'll have to agree to disagree. I think it is important to have centralized clean up, but I don't want to merge without consensus.

@davidfowl can you please take a look. Thanks.

**Comment by drewnoakes** [comment]
> I think it is confusing that the [resource publisher does this](https://github.com/dotnet/aspire/blob/5ee685ee232856f25de5ae59ba0325e2d8e11448/src/Aspire.Hosting/Dashboard/ResourcePublisher.cs#L47-L49), but the log publisher doesn't.

I'm not sure what you mean. The log publisher returns objects which do that. `ResourcePublisher` doesn't implement `IDisposable` either. I feel like I'm missing something.

**Comment by mitchdenny** [comment]
Give @davidfowl's refactoring in this area is this PR still useful?

**Comment by davidfowl** [comment]
No these are going to be deleted soon. Streaming logs will replace these classes. @karolz-ms is working on that

**Pull request closed without merging**